ambry-cloud/src/main/java/com/github/ambry/cloud/azure/ADAuthBasedStorageClient.java

```diff
@@ -87,4 +121,22 @@ private IAuthenticationResult getAccessTokenByClientCredentialGrant(AzureCloudCo
         ClientCredentialParameters.builder(Collections.singleton(azureCloudConfig.azureStorageScope)).build();
     return app.acquireToken(clientCredentialParam).get();
   }
+
+  @Override
+  protected boolean tryHandleExceptionAndHintRetry(BlobStorageException blobStorageException) {
+    // If the exception has status code 403, refresh the token and create a new storage client with the new token.
+    if (blobStorageException.getStatusCode() == HttpStatus.SC_FORBIDDEN) {
+      synchronized (this) {
+        // check if the access token has expired before refreshing the token. This is done to prevent multiple threads
+        // to attempt token refresh at the same time. It is expected that as a result of token refresh, accessTokenRef
+        // will updated with the new token.
+        if (accessTokenRef.get().isExpired()) {
+          BlobServiceClient blobServiceClient = createBlobStorageClient();
+          setClientReferences(blobServiceClient);
+        }
+      }
+      return true;
```

Comment:
If the token was not expired, should we still retry?  What if it's a non-transient permission issue?